[GEOT-5421] NullPointerException when GridCoverage goes offscreen

matchami opened the pull request:
Superceeds PR 1204 (in branch) -> master

bencaradocdavies commented on modules/library/render/src/test/java/org/geotools/renderer/lite/gridcoverage2d/GridCoverageRenderingOutOfViewTest.java:
```diff
@@ -0,0 +1,101 @@
+ *    GeoTools - The Open Source Java GIS Toolkit
+ *    http://geotools.org
+ * 
+ *    (C) 2016, Jumbletree Ltd
```
@matchami there is a problem with this copyright header. To contribute code to the GeoTools project, copyright must be assigned to OSGeo. Source code files must use the following header:
http://docs.geotools.org/latest/developer/introduction.html#source-license

As you are adding a new file, please submit a code contribution agreement. If this work is owned by Jumbletree Ltd, then use the corporate agreement signed by an authorised person:
http://docs.geotools.org/latest/developer/procedures/contribution_license.html

Please also copy @jodygarnett (OSGeo project officer) when you submit your agreement.

Sorry for not spotting this earlier.

I also urge you to squash these commits into a single commit for clarity. You can do this with ```git cherry-pick -n COMMIT_ID ...``` to prepare your changes and then combine them into a single commit.

I also see that you are working on master in your fork. One of the big changes in Git (from svn) is that branches are super-easy to create and use, and better than working on master. For example, it is easy to create a branch ```geot-5421-master```, make your changes as a single commit, push to your GitHub fork, and make a pull request. This allows you to work on your next issue while your pull request is outstanding.

matchami replied:
Hi @bencaradocdavies and @jodygarnett

Please find attached my signed contribution agreement.

Thanks

Iain


On 15/06/2016 10:43 a.m., Ben Caradoc-Davies wrote:
>
> In 
> modules/library/render/src/test/java/org/geotools/renderer/lite/gridcoverage2d/GridCoverageRenderingOutOfViewTest.java 
> <https://github.com/geotools/geotools/pull/1212#discussion_r67071493>:
>
> > @@ -0,0 +1,101 @@
> > +/*
> > + *    GeoTools - The Open Source Java GIS Toolkit
> > + *    http://geotools.org
> > + *
> > + *    (C) 2016, Jumbletree Ltd
>
> @matchami <https://github.com/matchami> there is a problem with this 
> copyright header. To contribute code to the GeoTools project, 
> copyright must be assigned to OSGeo. Source code files must use the 
> following header:
> http://docs.geotools.org/latest/developer/introduction.html#source-license
>
> As you are adding a new file, please submit a code contribution 
> agreement. If this work is owned by Jumbletree Ltd, then use the 
> corporate agreement signed by an authorised person:
> http://docs.geotools.org/latest/developer/procedures/contribution_license.html
>
> Please also copy @jodygarnett <https://github.com/jodygarnett> (OSGeo 
> project officer) when you submit your agreement.
>
> Sorry for not spotting this earlier.
>
> I also urge you to squash these commits into a single commit for 
> clarity. You can do this with |git cherry-pick -n COMMIT_ID ...| to 
> prepare your changes and then combine them into a single commit.
>
> I also see that you are working on master in your fork. One of the big 
> changes in Git (from svn) is that branches are super-easy to create 
> and use, and better than working on master. For example, it is easy to 
> create a branch |geot-5421-master|, make your changes as a single 
> commit, push to your GitHub fork, and make a pull request. This allows 
> you to work on your next issue while your pull request is outstanding.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/geotools/geotools/pull/1212/files/1bf3670edbfc34d2332bf1ec017f8eef0260a305#r67071493>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AChibFtTsbf5ewOn8UqnXbd959mQVNjyks5qLy6GgaJpZM4I10aG>.
>

-- 
Sent from my ZX80



---
This email has been checked for viruses by Avast antivirus software.
https://www.avast.com/antivirus

matchami commented:
Thanks @bencaradocdavies for the info, especially (again) around git as it is great to learn these things.  Next time I will do as you suggest and create a branch (I did wonder about what would happen with the 'next' issue :) and will also use -n to squash the commits into one

At the moment I am under managerial pressure to stop spending time on this.  We are 1000% happy to contribute to the open source project etc but when the actual code takes you half an hour to write but you're spending three or four hours now trying to jump through technicality hoops which could have been done in five minutes (eg if you had been the person to respond to my first PR instead of @aaime ), the management's enthusiasm level seems to drop quite rapidly for some reason ;)

With regards to the copyright it just didn't seem to make sense to me to put a copyright into the code for an organisation that I have nothing to do with.  Had Andrea been as helpful as you have been rather than just throwing one-line demands at the newbie (me) this could have been sorted sooner.  I have fixed now :)

I will discuss with management where the ownership actually lies fill in the appropriate form as you've asked.

For now, please would you be able to take over and do the squash into one commit if you think that is appropriate as I'll have trouble justifying more code-juggling to the boss.  A really big thank you again for your help.

bencaradocdavies commented:
@matchami that copyright header fix looks good (I will amend the copyright date to 2016!). I am happy to squash your commits into one and and apply on master as soon as we get a signed contributor agreement. Please copy @jodygarnett and me when you email it to OSGeo.

Software can be hard, but hardest of all are the non-software parts: people, governance, and community. Yes, there is an overhead, but paying attention to these is necessary for this project to endure.

Thanks again for your contribution. And thanks to @aaime for the code review.

matchami commented:
Sorry didn't spot the date...copy and paste is bad :).  Still waiting to 
hear back from mgmt re the agreement but I will get it off to you and 
Jody as soon as I get an answer.  Thanks!


On 15/06/2016 12:50 p.m., Ben Caradoc-Davies wrote:
>
> @matchami <https://github.com/matchami> that copyright header fix 
> looks good (I will amend the copyright date to 2016!). I am happy to 
> squash your commits into one and and apply on master as soon as we get 
> a signed contributor agreement. Please copy @jodygarnett 
> <https://github.com/jodygarnett> and me when you email it to OSGeo.
>
> Software can be hard, but hardest of all are the non-software parts: 
> people, governance, and community. Yes, there is an overhead, but 
> paying attention to these is necessary for this project to endure.
>
> Thanks again for your contribution. And thanks to @aaime 
> <https://github.com/aaime> for the code review.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/geotools/geotools/pull/1212#issuecomment-226059280>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AChibHl4t6-oz4ADFcz6MaJU7YiRWMA3ks5qL0xKgaJpZM4I10aG>.
>

-- 
Sent from my ZX80



---
This email has been checked for viruses by Avast antivirus software.
https://www.avast.com/antivirus

bencaradocdavies commented:
@matchami you also had tabs in your pom change and CRLF line endings in ```GridCoverageRenderingOutOfViewTest``` (please check your git line ending settings); I fixed these, ran test builds of GeoTools and GeoServer (which passed), and pushed the resulting squashed commit into a branch of my fork:
https://github.com/bencaradocdavies/geotools/commit/214b056db032a4ca652e3c136e75ef663e81ea3e

All ready to be cherry-picked onto master when we have your contributor agreement.

In this case I used a git squash merge to combine your commits into a feature branch. This has pretty much the same effect cherry-pick -n but is more straightforward when dealing with several commits.

matchami commented:
Thanks @bencaradocdavies 
<https://github.com/bencaradocdavies/geotools/commit/214b056db032a4ca652e3c136e75ef663e81ea3e> 
I have set my git autocrlf to true and have amended the editor settings 
for my GT projects to use spaces instead of tabs so that this should be 
fixed for the future.  The "next" one will be better, promise :)


On 15/06/2016 2:25 p.m., Ben Caradoc-Davies wrote:
>
> @matchami <https://github.com/matchami> you also had tabs in your pom 
> change and CRLF line endings in |GridCoverageRenderingOutOfViewTest| 
> (please check your git line ending settings); I fixed these, ran test 
> builds of GeoTools and GeoServer (which passed), and pushed the 
> resulting squashed commit into a branch of my fork:
> bencaradocdavies@214b056 
> <https://github.com/bencaradocdavies/geotools/commit/214b056db032a4ca652e3c136e75ef663e81ea3e>
>
> All ready to be cherry-picked onto master when we have your 
> contributor agreement.
>
> In this case I used a git squash merge to combine your commits into a 
> feature branch. This has pretty much the same effect cherry-pick -n 
> but is more straightforward when dealing with several commits.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/geotools/geotools/pull/1212#issuecomment-226071302>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AChibA4vO182GaXoK3hgAo7Cu8QlzUUjks5qL2KXgaJpZM4I10aG>.
>

-- 
Sent from my ZX80



---
This email has been checked for viruses by Avast antivirus software.
https://www.avast.com/antivirus

aaime commented:
"Had Andrea been as helpful as you have been rather than just throwing one-line demands at the newbie (me) this could have been sorted sooner. "

When you submit a pull request github presents a link to the contribution.md file and asks you to read it: https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md
I agree it's too easy to ignore, indeed I've contacted github about it, but no action has been taken so far.

Nobody here is paid to perform reviews, we do that in our spare time. When faced with your pull request the first time I had five minutes and scanned though it, the lack of a test was the first thing jumping at me, and just replied... in retrospect that was a mistake, I should have just pointed you at the contribution guidelines, or not have replied at all (I simply don't have the time anymore to be as thorough as @bencaradocdavies has been).

matchami commented:
Thanks @aaime.  Your comments were to the point and well received - but 
it would have been nice if you could have spared some time to respond to 
my requests for help regarding things that I just did not know how to 
do.  Even pointing me at this file that you have mentioned now would 
have saved me several hours, and probably you much frustration at me!

To be honest with you I have not seen the link to the contributing.md 
file even after creating three pull requests, so I would support your 
suggestion to github that it be made more prominent.

Thanks! Iain


On 15/06/2016 5:43 p.m., Andrea Aime wrote:
>
> "Had Andrea been as helpful as you have been rather than just throwing 
> one-line demands at the newbie (me) this could have been sorted sooner. "
>
> When you submit a pull request github presents a link to the 
> contribution.md file and asks you to read it: 
> https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md
> I agree it's too easy to ignore, indeed I've contacted github about 
> it, but no action has been taken so far.
>
> Nobody here is paid to perform reviews, we do that in our spare time. 
> When faced with your pull request the first time I had five minutes 
> and scanned though it, the lack of a test was the first thing jumping 
> at me, and just replied... in retrospect that was a mistake, I should 
> have just pointed you at the contribution guidelines, or not have 
> replied at all (I simply don't have the time anymore to be as thorough 
> as @bencaradocdavies <https://github.com/bencaradocdavies> has been).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/geotools/geotools/pull/1212#issuecomment-226093332>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AChibA1EIHVwHZ_oHTCDjK7Kahdu7-Dnks5qL5DugaJpZM4I10aG>.
>

-- 
Sent from my ZX80



---
This email has been checked for viruses by Avast antivirus software.
https://www.avast.com/antivirus

aaime commented:
Yeah, the link is right at the top of the page, but we are so used to ignoring this kind of requests (e.g., stuff like "read the licence") that most of us automatically phase it out:

![image](https://cloud.githubusercontent.com/assets/325433/16074488/fe1d2186-32ea-11e6-8568-2645ed0799c2.png)

bencaradocdavies commented:
Cherry-picked onto master: https://github.com/geotools/geotools/commit/f028e8370f82c8779dc881ac9c9dc4c49c6a00cb
